Add type integrity check when existing types are found

GeertvanHorrik opened the pull request:
This PR adds type integrity checks when a type is (accidentally) registered twice.

This fixes #3331. For example, in my case it was not applying the name conversion (camelCase) to the 2nd object causing runtime issues.

Example of an exception being thrown. This should give the developer some pointers to look for registration issues.

```
Type 'MyObjectGraphType' has field 'id', but this cannot be found on the new type. New type has:
- Id
- Name

Make sure to use the same type instances when registering the same type
```

Question: this PR causes 1 unit tests for a bug (Issue1927ResolvedTypeWithType, test_outputs) to fail. How would you like to deal with this?

Shane32 commented:
I’ll review in a bit. Ideally we would retain a test for whatever 1927 addressed, while fixing the issue that is interfering with the new patch. Feel free to edit the old test and I’ll review the changes.

Shane32 commented:
I’d like to have your test case moved to a Bug3331.cs file while adding a test within SchemaTypesTests that is very narrow and tests the exact scenario we are fixing. It should be possible within a dozen lines of code or so - just add two fields to a query type that each have their ResolvedType properties set to different instances of an object graph type with the same name. It should fail upon calling schema.Initialize

GeertvanHorrik commented:
All the methods on the SchemaTypes were private. I will keep the simple test on the SchemaTypes unit tests and move the complex one to the bug file. Does that sound good?

Shane32 commented:
Sounds good.

Shane32 commented on src/GraphQL/Types/Collections/SchemaTypes.cs:
```diff
@@ -646,11 +647,62 @@ private void AddTypeIfNotRegistered(IGraphType type, TypeCollectionContext conte
+            }
+
+            // Ignore scalars
+            if (existingType is not IObjectGraphType objectKnownType || newType is not IObjectGraphType objectNewType)
```
This check is good; scalars do not matter if multiple instances are being used, and it is likely enough that it happens throughout the codebase.  However, the same issue applies to `InterfaceGraphType` and `InputObjectGraphType` instances as well as `IObjectGraphType` instances

I'd probably just combine this with the `ReferenceEquals` check above but either way is fine.

Shane32 commented on src/GraphQL/Types/Collections/SchemaTypes.cs:
```diff
@@ -646,11 +647,62 @@ private void AddTypeIfNotRegistered(IGraphType type, TypeCollectionContext conte
+
+            if (existingType.GetType() != newType.GetType())
+            {
+                throw new InvalidOperationException($"Type '{existingType.GetType().Name}' is already registered but being re-registered using different type '{newType.GetType().Name}'");
```
I suggest we expand this error message a bit to include the GraphQL type name that is registered a second time.

Shane32 commented on src/GraphQL/Types/Collections/SchemaTypes.cs:
```diff
@@ -646,11 +647,62 @@ private void AddTypeIfNotRegistered(IGraphType type, TypeCollectionContext conte
+                return;
+            }
+
+            var knownFields = new HashSet<string>(objectKnownType.Fields.Select(x => x.Name));
+            var newFields = new HashSet<string>(objectNewType.Fields.Select(x => x.Name));
+
+            if (knownFields.Count != newFields.Count)
+            {
+                throw new InvalidOperationException($"Type '{existingType.GetType().Name}' has '{knownFields.Count}' field(s), where new type has '{newFields.Count}' field(s). Make sure to use the same type instances when registering the same type");
+            }
+
+            foreach (var knownTypeField in objectKnownType.Fields)
+            {
+                var newTypeField = objectNewType.Fields.FirstOrDefault(x => string.Equals(knownTypeField.Name, x.Name));
+                if (newTypeField is null)
+                {
+                    var exceptionStringBuilder = new StringBuilder($"Type '{existingType.GetType().Name}' has field '{knownTypeField.Name}', but this cannot be found on the new type. New type has:");
+                    exceptionStringBuilder.AppendLine();
+
+                    foreach (var newField in objectNewType.Fields)
+                    {
+                        exceptionStringBuilder.AppendLine($"- {newField.Name}");
+                    }
+
+                    exceptionStringBuilder.AppendLine();
+                    exceptionStringBuilder.Append("Make sure to use the same type instances when registering the same type");
+
+                    throw new InvalidOperationException(exceptionStringBuilder.ToString());
+                }
+            }
```
Let's replace all of this with a simple exception.  What if all fields match?  It should still generate an exception.

```suggestion
            throw new InvalidOperationException($"A different instance of the type '{newType.Name}' has already been registered within the schema. Please use the same instance for all references within the schema, or use GraphQLTypeReference to reference a type instantiated elsewhere.");
```

Shane32 commented:
It seems that the `Issue1927ResolvedTypeWithType` tests are easily fixed by adding a name to the inner type.

```cs
        var innerObject = new ObjectGraphType() { Name = "inner" };
```

However, it reveals an issue with `RegisterType` where registering the same type multiple times causes an issue -- probably the same issue you ran into.  See failing test `SchemaTests.registers_additional_duplicated_types` and `Bug1046.any_registration_order_should_work`, plus a bunch in `OverlappingFieldsCanBeMergedTest`.

I think we need to fix the underlying issue here, specifically that calling `RegisterType<T>` on a type that has already been registered (directly or indirectly) does not attempt to reregister the type.  Probably changing some call from `AddType` to `AddTypeIfNotRegistered` or similar within `SchemaTypes`.

GeertvanHorrik commented:
Thank you very much for the code review.

I've addressed all the issues. I tried adding the parent type info (because now it's still hard to pinpoint where it goes wrong), but it seems this context is not available in the context nor the types.

When changing the method to a more simplified version, it does cause more tests to fail (which I think is a good thing, this can lead to very hard to find issues). Will fix them once you approved my changes so far.

GeertvanHorrik commented:
> I think we need to fix the underlying issue here, specifically that calling RegisterType<T> on a type that has already been registered (directly or indirectly) does not attempt to reregister the type. Probably changing some call from AddType to AddTypeIfNotRegistered or similar within SchemaTypes.

Made the change to the calls (except where I think it didn't make much sense). There are now 23 failing tests, showing potential issues.

I think the parents (of both old and new) should be passed into the AddTypeIfNotRegistered. This will give more context. Would that be a good change?

Shane32 commented on src/GraphQL/Types/Collections/SchemaTypes.cs:
```diff
@@ -646,11 +646,37 @@ private void AddTypeIfNotRegistered(IGraphType type, TypeCollectionContext conte
+
+            if (existingType.GetType() != newType.GetType())
+            {
+                throw new InvalidOperationException($@"Unable to register GraphType '{newType.GetType().FullName}' with the name '{newType.Name}'. The name '{newType.Name}' is already registered to '{existingType.GetType().FullName}'. Check your schema configuration.");
```
We have a `GetFriendlyName()` extension method for `Type`s rather than `FullName` which we typically use as a matter of convention.
```suggestion
                throw new InvalidOperationException($@"Unable to register GraphType '{newType.GetType().GetFriendlyName()}' with the name '{newType.Name}'. The name '{newType.Name}' is already registered to '{existingType.GetType().GetFriendlyName()}'. Check your schema configuration.");
```

GeertvanHorrik replied:
I copied this from another location. Will change it there as well.

Shane32 commented on src/GraphQL.Tests/Types/Collections/SchemaTypesTests.cs:
```diff
@@ -45,6 +46,46 @@ public void does_not_request_instance_more_than_once()
+        queryGraphType.Field("second", graphType2);
+
+        // Test
+        Assert.Throws<InvalidOperationException>(() => schema.Initialize());
```
We use the Shouldly assertion library for tests, and as a matter of convention, test the message name (more so it can be seen what the failure is rather than because we want to ensure it does not change).

```suggestion
        Should.Throw<InvalidOperationException>(() => schema.Initialize())
            .Message.ShouldBe("...");
```

GeertvanHorrik commented:
Thanks for the review. Points should be addressed. Only my previous comment remains: how to deal with the other unit tests?

Shane32 commented on src/GraphQL/Types/Collections/SchemaTypes.cs:
```diff
@@ -189,7 +189,7 @@ protected void Initialize(ISchema schema, IServiceProvider serviceProvider, IEnu
                     if (this[name] == null)
                     {
-                        AddType(graphType, context);
+                        AddTypeIfNotRegistered(graphType, context);
```
Well, unfortunately this isn't helping.  I did a little investigation and found that `GetSchemaTypes` is going to create instances for each of the types listed in `AdditionalTypes`.  That could be solved easily enough with `.Distinct()` but it does not cover the situation where an indirect reference to a type previously added via `RegisterType` has a separate reference from the subsequent one added via `RegisterType`.  I don't have time to look at it further at the moment.

Shane32 replied:
So might as well revert this line at this time.

Shane32 commented on src/GraphQL/Types/Collections/SchemaTypes.cs:
```diff
@@ -603,7 +603,7 @@ private void AddTypeWithLoopCheck(IGraphType resolvedType, TypeCollectionContext
             try
             {
-                AddType(resolvedType, context);
+                AddTypeIfNotRegistered(resolvedType, context);
```
Here either.  Might as well revert.

Shane32 commented:
We could probably make line 201 to pass a boolean to `AddTypeIfNotRegistered` so that it does not throw an exception if a duplicate is encountered.  Theoretically this should be fine, but this feels like a hack and not a solution.  For example, then it would then not throw an exception if two different instances were manually registered.

Shane32 commented:
(At least the schema would be consistent)

Shane32 commented:
Really, the types in `AdditionalTypes` should be calling the other overload of `AddTypeIfNotRegistered` that accepts a `Type` and not a `IGraphType`.

Shane32 commented:
I suggest:

```cs

        private static (IEnumerable<IGraphType>, IEnumerable<Type>) GetSchemaTypes(ISchema schema, IServiceProvider serviceProvider)
        {
            return (
                GetSchemaTypeInstances(),
                schema.AdditionalTypes.Select(x => x.GetNamedType()).Where(x => !typeof(ScalarGraphType).IsAssignableFrom(x)));

            IEnumerable<IGraphType> GetSchemaTypeInstances()
            {
                // Manually registered AdditionalTypeInstances and AdditionalTypes should be handled first.
                // This is necessary for the correct processing of overridden built-in scalars.

                foreach (var instance in schema.AdditionalTypeInstances)
                    yield return instance;

                foreach (var type in schema.AdditionalTypes)
                {
                    var type2 = type.GetNamedType();
                    if (typeof(ScalarGraphType).IsAssignableFrom(type2))
                    {
                        yield return (IGraphType)serviceProvider.GetRequiredService(type2);
                    }
                }

                //TODO: According to the specification, Query is a required type. But if you uncomment these lines, then the mass of tests begin to fail, because they do not set Query.
                // if (Query == null)
                //    throw new InvalidOperationException("Query root type must be provided. See https://spec.graphql.org/October2021/#sec-Schema-Introspection");

                if (schema.Query != null)
                    yield return schema.Query;

                if (schema.Mutation != null)
                    yield return schema.Mutation;

                if (schema.Subscription != null)
                    yield return schema.Subscription;
            }
        }

//line 147

            var (typeinstances, types) = GetSchemaTypes(schema, serviceProvider);

//line 169

            // Add manually-added scalar types. To allow overriding of built-in scalars, these must be added
            // prior to adding any other types (including introspection types).
            foreach (var type in typeinstances)
            {
                if (type is ScalarGraphType)
                    AddType(type, _context);
            }

//line 198

            foreach (var type in typeinstances)
            {
                if (type is not ScalarGraphType)
                    AddTypeIfNotRegistered(type, ctx);
            }

            foreach (var type in types)
            {
                AddTypeIfNotRegistered(type, ctx);
            }
```

I believe that will fix all of the other tests

Shane32 commented on src/GraphQL.Tests/Bugs/Bug3331.cs:
```diff
@@ -0,0 +1,175 @@
+            .Message.ShouldBe("A different instance of the type 'MyObject' has already been registered within the schema. Please use the same instance for all references within the schema, or use GraphQLTypeReference to reference a type instantiated elsewhere.");
+
+        // Must have 2 instances
+        Assert.Equal(3, MyObjectGraphType.SharedInstanceCounter);
```
```suggestion
        MyObjectGraphType.SharedInstanceCounter.ShouldBe(3);
```

Shane32 commented:
Have to change line 240-241 also to this:

```cs
                foreach (var type in schema.AdditionalTypes)
                {
                    var type2 = type.GetNamedType();
                    if (typeof(ScalarGraphType).IsAssignableFrom(type2))
                    {
                        yield return (IGraphType)serviceProvider.GetRequiredService(type2);
                    }
                }
```

And then all tests should pass.

GeertvanHorrik commented:
That's it, everything seems to run fine now. Thanks for all the review work.

If you want anything changed, please let me know and I'll try to fix asap.

Shane32 commented:
I think this is great; thanks for your help!
